pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,99 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.RWMutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	sc.lock.RLock()
+	defer sc.lock.RUnlock()
+	selector, ok := sc.cache[key]
+	return selector, ok
+}
+
+// Update can update selector of ServiceSelectorCache while service's selector changed.
+// Update can also add selector to ServiceSelectorCache while service's selector no cached.
+func (sc *ServiceSelectorCache) Update(key string, service *v1.Service) labels.Selector {
```

Comment:
The only attribute in use here from the Service is the selector, it might be simpler to accept that as an argument instead of the full service. That would also simplify testing below.